lua/nvim-surround/config.lua

```diff
@@ -24,77 +14,474 @@ M.default_opts = {
+---@return selection? The retrieved selection.
+M.get_selection = function(args)
+    if args.pattern then
+        return require("nvim-surround.patterns").get_selection(args.pattern)
+    elseif args.textobject then
+        return require("nvim-surround.textobjects").get_selection(args.textobject)
+    end
+end
+
+-- Gets a pair of selections from the buffer based on some heuristic.
+---@param args { char: string?, pattern: string? }
+M.get_selections = function(args)
+    if args.char and args.pattern then
+        return require("nvim-surround.utils").get_selections(args.char, args.pattern)
+    end
+end
+
+--[====================================================================================================================[
+                                                End of Helper Functions
+--]====================================================================================================================]
+
 -- Stores the global user-set options for the plugin.
-M.user_opts = {}
+M.user_opts = nil
 
 -- Returns the buffer-local options for the plugin, or global options if buffer-local does not exist.
 ---@return options @The buffer-local options.
 M.get_opts = function()
     return vim.b[0].nvim_surround_buffer_opts or M.user_opts
 end
 
+-- Returns the add key for the surround associated with a given character, if one exists.
+---@param char string? The input character.
+---@return fun(string?): string[][]? @The function to get the delimiters to be added.
+M.get_add = function(char)
+    char = require("nvim-surround.utils").get_alias(char)
+    local key = M.get_opts().delimiters[char] or M.get_opts().delimiters.invalid_key_behavior
+    return key.add
+end
+
+-- Returns the delete key for the surround associated with a given character, if one exists.
+---@param char string? The input character.
+---@return fun(string?): selections? @The function to get the selections to be deleted.
+M.get_delete = function(char)
+    char = require("nvim-surround.utils").get_alias(char)
+    local key = M.get_opts().delimiters[char] or M.get_opts().delimiters.invalid_key_behavior
+    return key.delete
+end
+
+-- Returns the change key for the surround associated with a given character, if one exists.
+---@param char string? The input character.
+---@return { target: function, replacement: function? }? @A table holding the target/replacment functions.
+M.get_change = function(char)
+    char = require("nvim-surround.utils").get_alias(char)
+    local key = M.get_opts().delimiters[char] or M.get_opts().delimiters.invalid_key_behavior
+    return key.change
+end
+
+-- Translates the user-provided configuration into the internal form.
+---@param opts options? The user-provided options.
+M.translate_opts = function(opts)
+    if not (opts and opts.delimiters) then
+        return opts
+    end
+    local invalid = opts.delimiters.invalid_key_behavior or M.default_opts.delimiters.invalid_key_behavior
```

Comment:
Could you maybe add another deprecation warning for `invalid_key_behavior`? (there was an error for me because I still had it set to a function). Something like this:
```lua
if type(invalid) == "function" then
    local invalid_key_warning = {
        "The old way of configuring `invalid_key_behavior` has been deprecated;",
        "see :h nvim-surround.config.delimiters for details.",
    }
    vim.notify_once(table.concat(invalid_key_warning, "\n"), vim.log.levels.WARN)
end
```